### PR TITLE
LF: Remove unecessary logic to pull package dependencies

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
in the preprocessor and the validation, since, the compiler is already doing that.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala [hunk 1]
```diff
@@ -222,21 +189,19 @@ private[preprocessing] object Preprocessor {
   }
 
   @inline
-  def safelyRun[X](
-      handleMissingPackages: Result[_]
-  )(unsafeRun: => X): Result[X] = {
+  def safelyRun[X](handleMissingPackages: => Result[_])(unsafeRun: => X): Result[X] = {
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:192]: `handleMissingPackages` should have always been called by name.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala [hunk 2]
```diff
@@ -51,127 +51,96 @@ private[engine] final class Preprocessor(
...
-    safelyRun(getDependencies(List(ty0), List.empty)) {
+    safelyRun(pullTypePackages(ty0)) {
       commandPreprocessor.valueTranslator.unsafeTranslateValue(ty0, v0)
     }
 
   private[engine] def preprocessApiCommand(
       cmd: command.ApiCommand
   ): Result[speedy.Command] =
-    safelyRun(getDependencies(List.empty, List(cmd.templateId))) {
+    safelyRun(pullTemplatePackage(List(cmd.templateId))) {
```
Comment (nickchapman-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:143]: So my understanding is that the essence of this change is to split the old code `getDependencies`
into two parts `pullTypePackage` and `pullTemplatePackage`
which pull packages based on references to either template of type refs.

Does this this avoid some work, or is it simply a cleaner design?

**Review by nickchapman-da — APPROVED** [review]
I had some questions, but LGTM !

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:192]: Yes.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:143]: We avoid indeed some work. 
Before we were pulling the definitions of the type constructors and the templates we encountered and we were processing recursively all the types appearing in those definition (previous lines 93 to 111 and from 135 to 147). Now we avoid completely this.

**Pull request merged**